### PR TITLE
Make Strapi GraphQL an implementing service for Federation

### LINKED ISSUE #4870: GraphQL Federation (schema stitching)
- [x] **I have created my request on the Product Board before I submitted this issue**
- [x] **I have looked at all the other requests on the Product Board before I submitted this issue**

**Please describe your feature request:**

This would allow for easy schema stitching with other GraphQL services. See specs [here](https://www.apollographql.com/docs/apollo-server/federation/introduction/).

### PR DESCRIPTION
**Pull request by shaheinm**
#### Description of what you did:
Fix #4870
Added `federation` flag to GraphQL plugin config.
- If disabled, works exactly as before.
- If enabled, the `Upload` scalar created by Strapi is **not** removed, as a Federated schema requires the use of `@apollo/federation` package's `buildFederatedSchema()` function, and that function uses `makeExecutableSchema` under the hood, requiring the inclusion of the scalar `Upload`.

**NOTE**: I purposely did not include anything related to the `@key` directive, which sets up the ability to extend types in other services included in Federation. You can use Strapi's custom GraphQL schema definition to do this, though.

No tests were added or changed, but all are passing. Documentation was updated accordingly.

### CODE REVIEW

File: packages/strapi-plugin-graphql/services/schema-generator.js [hunk 1]
```diff
@@ -90,8 +90,10 @@ const generateSchema = () => {
     writeGenerateSchema(graphql.printSchema(schema));
   }
 
-  // Remove custom scalar (like Upload);
-  typeDefs = Types.removeCustomScalar(typeDefs, resolvers);
+  // Remove custom scalar (like Upload), if not using Federation
+  if (!strapi.plugins.graphql.config.federation) {
```
Comment (alexandrebodin) [inline packages/strapi-plugin-graphql/services/schema-generator.js:94]: Can you explain why this is only need when not using federation ? 

Also can you use _.get tu be consistent :)

Comment (shaheinm) [reply packages/strapi-plugin-graphql/services/schema-generator.js:94]: Using Federation requires the use of `buildFederatedSchema` when creating the graphql server, which means the scalar Upload needs to be defined in the schema for file uploads to still work.

I'll make the consistency change and update the comment to be a bit clearer later today.

**Comment by shaheinm** [comment]
Any updates on this @alexandrebodin

**Comment by alexandrebodin** [comment]
Hey @shaheinm can you give more context / example as to what you can do with this added ?

**Comment by bamne123** [comment]
When it will be in Release?

**Comment by shaheinm** [comment]
> Hey @shaheinm can you give more context / example as to what you can do with this added ?

@alexandrebodin One example is simply running Strapi as a microservice with Apollo Federation. My team is planning to customize the Strapi GraphQL schema to allow it to be extended in a versioning and indexing service, which will act as a sort of single source of truth, while Strapi provides the data on the client. We're also handling file uploads outside of Strapi, and simply plugging links into the content, and wanted that service at the same url for the client side application to access.

**Comment by efstajas** [comment]
👀 Would love to see this. We're also running a microservice environment with Federation and having our blog content alongside other service's data would be sweet. Please let me know if we can assist with getting this merged in any way.

**Comment by alexandrebodin** [comment]
Right now the PR has conflicts. I would like to see if we can avoid having the dependency installed for everyone though 🤔  using some customization to add it programmaticaly maybe

**Comment by ntsvetkova-fcse** [comment]
We also would like to use Federation instead of creating a Strapi connector in our middleware. Would be really nice to have this PR merged ASAP.

**Comment by shaheinm** [comment]
> Right now the PR has conflicts. I would like to see if we can avoid having the dependency installed for everyone though  using some customization to add it programmaticaly maybe

I can try to resolve the conflicts in yarn.lock but I don't believe there is any way to avoid adding the dependency in package.json. Moreover, the `@apollo/federation` package is only 606 kb unpacked, so it isn't a major addition.

**Comment by alexandrebodin** [comment]
@lauriejim Can you review the doc section ?

**Comment by maxbause** [comment]
We would also like to use the federation feature since it will allow us to "glue" multiple GraphQL servers (strapi, service 1, service 2, etc...) together which would be really handy. 💯

**Comment by benoj** [comment]
Hi,

I saw this was merged... but can't seem to see these changes in the code / docs (3.5.0) did this get removed?

**Comment by benoj** [comment]
in-fact - i have seen the changes to the config locations...

However when I run an apollo federation gateway I see the following:

```
Error checking for changes to service definitions: Couldn't load service definitions for "shop" at http://localhost:1337/graphql: 400: Bad Request
This data graph is missing a valid configuration. Couldn't load service definitions for "shop" at http://localhost:1337/graphql: 400: Bad Request

```

**Comment by AdrianOrlow** [comment]
@benoj It wasn't removed, but someone [changed flag name](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-graphql/services/schema-generator.js#L27) from `federation` to `isFederated`

**Comment by alexandrebodin** [comment]
Nice catch I'm going to make a fix for this but you can use isFederated as a workaround for now.

**Comment by derrickmehaffy** [comment]
This pull request has been mentioned on **Strapi Community Forum**. There might be relevant details there:

https://forum.strapi.io/t/is-it-possible-to-use-the-apollo-federation-specification/4334/3

**Comment by Tommo56700** [comment]
@alexandrebodin the documentation around this for Strapi v4 seems to be lacking. Are there any resources for how to set this up? specifically relating to the @key directive, which sets up the ability to extend types in other services included in Federation?

**Comment by danielsheard0** [comment]
This doesn't seem to be implemented in v4 unfortunately - would be great if it was 👀

**Pull request merged**